### ISSUE TITLE
I want the same feature in CMake: variable_watch

### ISSUE THREAD

**Issue by Jason2013** [mesonbuild/meson#9535]
I transfer from CMake to Meson and accept that Meson has no user defined functions and uses only global variables.

However, I cannot find the similar CMake `variable_watch` function in Meson. I have to spread `warning` function calls in the meson.build files and try to find when and where a variable is used or changed. It is boring and valueless work.

I think it is not hard to add the similar CMake `variable_watch`  function in Meson. It is just to add some logic to read and write operations of the global variable table.

variable_watch
https://cmake.org/cmake/help/latest/command/variable_watch.html

**Comment by eli-schwartz** [comment]
I don't understand what the objective of such functionality is. How do you use variables without triggering a warning when you assign to them for the first time? Or append to them with the `+=` operator which isn't exactly surprising that it modifies things?

I can't think of any language that does this (other than apparently cmake) and that includes turing-complete programming languages in addition to build system DSLs.

What are you trying to accomplish that you believe such a function could help with?

**Comment by Jason2013** [comment]
The `variable_wath` function is used for debug purpose.

The usage example is the following:

`variable_watch(<variable_name>)`

After this function call, all read and write operations to the `<variable_name>` variable will give a message including the following information:
1. the location of the read or write operation, in the format: file_name:line_number
2. the operation type: read or write
3. the current value of the variable and the old value of the variable


For example: If you have a very complex meson.build file and there are many global variables are used. You want to find where and when a specific variable is read or written and what value the variable is, then the `variable_watch` function can help a lot. Without the `variable_watch` function, I have to print the value of the variable by using `warning` function at the position that I think it might be modified.

**Comment by Jason2013** [comment]
For example: When you read a big meson.build file and find a variable is used at a line, how to find the current value of the variable? The simple way is to print the value of the variable just before the line like this: `warning(<variable_name>)`. If the value of the variable is not as your expected, you want to know: where the wrong value is assigned to the variable? The variable was assigned more than once. So you have to use `warning(<variable_name>)` after every line that the variable is assigned to find the value of the variable just after assignment operation. So that you will use `warning` function many times.

**Comment by eli-schwartz** [comment]
It sounds like what you really want is a debug mode of meson that logs some introspectable trace of actions taken, rather than adding a function to the meson DSL.

**Comment by Jason2013** [comment]
Yes. I am new to Meson and don't know how to use debug mode. Could you give some guides or page links?

**Comment by eli-schwartz** [comment]
Sadly, it doesn't exist yet. ;)

However, I propose that implementing such a thing is closer to the spirit of your desire than literally adding a `variable_watch` function. I think it's a reasonable thing to ask for.

**Comment by Jason2013** [comment]
That is great! If you have plan, please let me know.
As you known, there is no an easy tool for debugging CMake script but `variable_watch` function helps me a lot. I hope Meson has a similar feature to debug the meson script.
